app/controllers/application_controller.rb

```diff
@@ -35,8 +35,17 @@ class MissingParameterError < Exception; end
 
   def set_language
     set_gettext_locale
-    unless LANGUAGES.include? FastGettext.locale
-      params[:locale] = FastGettext.default_locale
+    requested_locale = FastGettext.locale
+    unless LANGUAGES.include? requested_locale
+      # XXX: if the browser sent de_DE, try de
+      l = requested_locale.split('_')
+      if LANGUAGES.include?l[0]
+        params[:locale] = l[0]
+      else
+        # not sure we need this. Gettext handles missing locale
+        # just fine
+        params[:locale] = FastGettext.default_locale
+      end
       set_gettext_locale
```

Comment:
As I'm not into ruby I didn't know about .first :-) The code is not the same though as we first need to check the variant with underscore (eg Portuguese and Chinese), then fallback to the one without underscore. Also, the fallback to default_locale is missing (although I think it's not needed.)